Dockerfile

```diff
@@ -1,24 +1,31 @@
-FROM ruby:2.3.1
+FROM ruby:2.3.1-alpine
 
-ENV RAILS_ENV=production
-ENV NODE_ENV=production
-
-RUN echo 'deb http://httpredir.debian.org/debian jessie-backports main contrib non-free' >> /etc/apt/sources.list
-RUN curl -sL https://deb.nodesource.com/setup_4.x | bash -
-RUN apt-get update -qq && apt-get install -y build-essential libpq-dev libxml2-dev libxslt1-dev nodejs ffmpeg && rm -rf /var/lib/apt/lists/*
-RUN npm install -g npm@3 && npm install -g yarn
-RUN mkdir /mastodon
+ENV RAILS_ENV=production \
+    NODE_ENV=production
 
 WORKDIR /mastodon
 
-ADD Gemfile /mastodon/Gemfile
-ADD Gemfile.lock /mastodon/Gemfile.lock
-RUN bundle install --deployment --without test development
-
-ADD package.json /mastodon/package.json
-ADD yarn.lock /mastodon/yarn.lock
-RUN yarn
+COPY . /mastodon
 
-ADD . /mastodon
+RUN BUILD_DEPS=" \
+    postgresql-dev \
+    libxml2-dev \
+    libxslt-dev \
+    build-base" \
+ && apk -U upgrade && apk add \
+    $BUILD_DEPS \
+    nodejs \
+    libpq \
+    libxml2 \
+    libxslt \
+    ffmpeg \
+    file \
+    imagemagick \
+ && npm install -g npm@3 && npm install -g yarn \
+ && bundle install --deployment --without test development \
```

Comment:
I'm messing with Docker again and I noticed this is suboptimal. While this Dockerfile optimizes for smallest possible end size, having it setup this way means that every time *anything* in the source directory changes, an entirely new image is created, all of these dependencies are installed from scratch etc. It not only takes a long time, but also means you end up with more needlessly different images on your filesystem which in turn eats up loads of disk space.

Dockerfiles must be structured this way:

1. Things that never change, e.g. build deps, system libraries/packages, users, system settings, etc
2. Dependency lockfiles, but *only* lockfiles, because we only want to trigger 3 to re-run if they change and nothing else
3. Installation of dependencies based on lockfiles
4. The rest of application code, which changes more often but requires no reinstallations of anything